src/olympia/devhub/views.py

```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
+                          'devhub.submit.version.finish')
+
+
+@dev_required
+@waffle_switch('step-version-upload')
+def submit_version(request, addon_id, addon):
+    if waffle.switch_is_active('mixed-listed-unlisted'):
+        # choose the channel we need from the last upload
+        last_version = addon.find_latest_version_including_rejected()
+        if not last_version:
+            return redirect('devhub.submit.version.distribution', addon.slug)
+        channel = last_version.channel
+    else:
+        # Don't allow channel choice until rest of AMO supports it.
+        channel = (amo.RELEASE_CHANNEL_LISTED if addon.is_listed else
+                   amo.RELEASE_CHANNEL_UNLISTED)
+    return _submit_upload(request, addon, channel,
+                          'devhub.submit.version.details',
+                          'devhub.submit.version.finish')
+
+
+def _submit_details(request, addon, version):
+    forms_list, context = [], {}
+    if version and version.channel == amo.RELEASE_CHANNEL_UNLISTED:
+        # Not a listed version ? Then nothing to do here.
+        return redirect('devhub.submit.version.finish', addon.slug, version.pk)
     # Figure out the latest version early in order to pass the same instance to
     # each form that needs it (otherwise they might overwrite each other).
-    latest_version = addon.find_latest_version(
+    latest_version = version or addon.find_latest_version(
         channel=amo.RELEASE_CHANNEL_LISTED)
     if not latest_version:
         # No listed version ? Then nothing to do in the listed submission flow.
         return redirect('devhub.submit.finish', addon.slug)
     post_data = request.POST if request.method == 'POST' else None
-
-    describe_form = forms.DescribeForm(
-        post_data, instance=addon, request=request)
-    cat_form = addon_forms.CategoryFormSet(
-        request.POST or None, addon=addon, request=request)
-    license_form = forms.LicenseForm(post_data, version=latest_version)
-    policy_form = forms.PolicyForm(post_data, addon=addon)
-    reviewer_form = forms.ReviewerNotesForm(
+    show_all_fields = not version or not addon.has_complete_metadata()
+
+    if show_all_fields:
+        describe_form = forms.DescribeForm(
+            post_data, instance=addon, request=request)
+        cat_form = addon_forms.CategoryFormSet(
+            post_data, addon=addon, request=request)
+        license_form = forms.LicenseForm(
+            post_data, version=latest_version, prefix='license')
+        context.update(license_form.get_context())
+        context.update(form=describe_form, cat_form=cat_form)
+        forms_list.extend([describe_form, cat_form, context['license_form']])
+    reviewer_form = forms.VersionForm(
         post_data, instance=latest_version)
+    context.update(reviewer_form=reviewer_form)
+    forms_list.extend([reviewer_form])
+
+    if request.method == 'POST' and all(
+            [form.is_valid() for form in forms_list]):
```

Comment:
You shouldn't need the `[]`, you can just do `all(form.is_valid() for form in forms_list)`